File: 1285-balance-a-binary-search-tree/balance-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void inorder(TreeNode* root,auto& v){
        if(!root) return;
        inorder(root->left,v);
        v.push_back(root->val);
        inorder(root->right,v);
    }
    TreeNode* balance(int l,int r,auto& v){
        if(l>r) return nullptr;
        int mid=l+(r-l)/2;
        TreeNode* root=new TreeNode(v[mid]);
        root->left=balance(l,mid-1,v);
        root->right=balance(mid+1,r,v);
        return root;
    }
    TreeNode* balanceBST(TreeNode* root) {
        vector<int> nodes;
        inorder(root,nodes);
        return balance(0,nodes.size()-1,nodes);
    }
};

File: 1285-balance-a-binary-search-tree/balance-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    allValues=[]
    def balanceBST(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        self.allValues.clear()
        self.inOrder(root)
        self.allValues.sort()
        n=len(self.allValues)
        return self.createBST(0,n-1)
    
    def createBST(self,l,r):
        if l>r:
            return None
        if l==r:
            return TreeNode(self.allValues[l])
        mid=l+(r-l)//2
        cur=TreeNode(self.allValues[mid])
        cur.left=self.createBST(l,mid-1)
        cur.right=self.createBST(mid+1,r)
        return cur


    def inOrder(self,root):
        if not root:
            return
        self.inOrder(root.left)
        self.allValues.append(root.val)
        self.inOrder(root.right)